keps/sig-api-machinery/3285-subresource-finalizers/README.md

```diff
@@ -0,0 +1,874 @@
+
+```
+<<[UNRESOLVED Endpoint name]>>
+Endpoint name for the subresource is still being discussed.
+<<[/UNRESOLVED]>>
+```
+
+### API Changes
+
+A new `Metadata` resource is introduced in `subresources.k8s.io` API group:
+
+```golang
+// Metadata represents a metadata field to be accessed via an endpoint for the subresource.
+type Metadata struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Only following fields should be set:
+	// - Namespace
+	// - Name
+	// - UID
+	// - ResourceVersion
+	// - The field for the endpoint, like Finalizers field for the finalizers endpoint
+	// If other fields are set, they are ignored.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+}
+```
+
+```
+<<[UNRESOLVED Resource name]>>
+Resource name is still being discussed.
+`Metadata` seems good if the same type can be used for all of the metadata subresources.
+If resource should be defined for each metadata subresource, `Finalize` or `Finalizers` will be candidates.
+<<[/UNRESOLVED]>>
+```
+
+### Supported API calls to the endpoint
+
+The `<resource>/finalizers` endpoint supports 3 verbs, GET, PATCH, and PUT.
+
+#### GET
+
+GET verb is used to read the value in `metadata.finalizers` field.
+
+- Request format:
+    Empty string should be passed as a request body. Any strings passed in the request body are ignored.
+
+- Response format:
+  - On success, the `Metadata` resource is returned. `Namespace`, `Name`, `UID`, `ResourceVersion`, and `Finalizers` fields are copied from the original resources.
+
+         ex)
+         ```json
+         {
+             "kind": "Metadata",
+             "apiVersion": "v1alpha1",
+             "metadata": {
+                 "namespace": "default",
+                 "name": "my-config",
+                 "uid": "1af42023-9303-4dd3-9383-f117fbd25296",
+                 "resourceVersion": "486",
+                 "finalizers": ["example.com/my-finalizer"],
+             }
+         }
+         ```
+
+  - On failure, error message is stored in `Status` resource.
+
+        ex)
+        ```json
+        {
+            "kind": "Status",
+            "apiVersion": "v1",
+            "metadata": {},
+            "status": "Failure",
+            "message": "\"finalizers\" is forbidden: User \"example-user\" cannot get subresource \"finalizers\" of \"ConfigMap\" in API group \"\" in the namespace \"default\"",
+            "reason": "Forbidden",
+            "details": {},
+            "code": 403
+        }
+        ```
+
+- Status code:
+  - 200 "OK": Returned on success without change
+  - 403 "Forbidden": Returned when GET verb is not allowed for `<resource>/finalizers` endpoint (RBAC admission handles this)
+
+#### PATCH
```

Comment:
I spoke with @apelisse on this and he said "there's no great answer for PATCH" and " it's mostly a footgun".

How about this?  Let's remove PATCH entirely and leave an UNRESOLVED for "Decide if we need PATCH support and if so, what the semantics are."

Users who want to add/remove a finalizer will have to RMW and use an RV precondition.

I'd rather us figure out how to allow more granular PUT/DELETE and justr not support PATCH at all.